include/aws/common/math.gcc_arm64_asm.inl

```diff
@@ -0,0 +1,184 @@
+#ifndef AWS_COMMON_MATH_GCC_ARM64_ASM_INL
+#define AWS_COMMON_MATH_GCC_ARM64_ASM_INL
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+/*
+ * This header is already included, but include it again to make editor
+ * highlighting happier.
+ */
+#include <aws/common/common.h>
+#include <aws/common/math.h>
+
+/* clang-format off */
+
+AWS_EXTERN_C_BEGIN
+
+/**
+ * Multiplies a * b. If the result overflows, returns 2^64 - 1.
+ */
+AWS_STATIC_IMPL uint64_t aws_mul_u64_saturating(uint64_t a, uint64_t b) {
+    /* We can use inline assembly to do this efficiently on arm64 by doing
+       a high-mul and checking  the upper 64 bits of a 64x64->128b multiply
+       are zero */
+    uint64_t tmp = 0, res = 0;
+    __asm__("umulh %x[hmul], %x[arga], %x[argb]\n"
+            "mul %x[res], %x[arga], %x[argb]\n"
+            "cmp %x[hmul], #0\n"
+            "csinv %x[res], %x[res], xzr, eq\n"
+            : /* inout: hmul is upper 64b, r is the result */ [hmul] "+&r"(tmp), [res]"+&r"(res)
+            : /* in: a and b */ [arga] "r"(a), [argb] "r"(b)
+            : /* clobbers: cc (cmp clobbers condition codes) */ "cc");
+
+    return res;
+}
+
+/**
+ * If a * b overflows, returns AWS_OP_ERR; otherwise multiplies
+ * a * b, returns the result in *r, and returns AWS_OP_SUCCESS.
+ */
+AWS_STATIC_IMPL int aws_mul_u64_checked(uint64_t a, uint64_t b, uint64_t *r) {
+    /* We can use inline assembly to do this efficiently on arm64 by doing
+       a high-mul and checking  the upper 64 bits of a 64x64->128b multiply
+       are zero */
+
+    uint64_t tmp, res;
+    __asm__("umulh %x[hmul], %x[arga], %x[argb]\n"
```

Comment:
do these __asm__ blocks need to be volatile?

Reply:
i don't see why they aren't modifying any state beyond what is marked as an output and if the compiler determines that the value isn't consumed it can be elided